### PR TITLE
RGW|BN: fix lifecycle test issue

### PR DESCRIPTION
**Pull request by AliMasarweh**
By polling on the bucket objects until empty, we make sure that we don't check for notification until we are sure that we should be receiving notifications about objects deletion

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by adamemerson** [comment]
jenkins test make check

File: src/test/rgw/bucket_notification/test_bn.py [hunk 1]
```diff
@@ -183,7 +183,8 @@ def do_POST(self):
             assert_equal(event['subject'], record['s3']['object']['key'])
             assert_equal(parser.parse(event['time']), parser.parse(record['eventTime']))
         log.info('HTTP Server received event: %s', str(body))
-        self.server.append(json.loads(body))
+        if body != b'':
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:186]: why do we have an empty message? and how can this happen if content length is not zero (see line 169)?

File: src/test/rgw/bucket_notification/test_bn.py [hunk 2]
```diff
@@ -1795,8 +1796,9 @@ def lifecycle(endpoint_type, conn, number_of_objects, topic_events, create_threa
     time.sleep(20)
     print('wait for sometime for the messages...')
 
-    no_keys = list(bucket.list())
     wait_for_queue_to_drain(topic_name, http_port=port)
+    time.sleep(20)
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1800]: does lifecycle takes more than 20 seconds?
if it does, why not wait longet in line 1796?
also, if queue is drained, it means we sent all notifications, why wait more?

Comment (yuvalif) [reply src/test/rgw/bucket_notification/test_bn.py:1800]: please loop over the bucket list until it ts empty. do a sleep (maybe 5s) inside the loop

File: src/test/rgw/bucket_notification/test_bn.py [hunk 3]
```diff
@@ -1791,11 +1791,15 @@ def lifecycle(endpoint_type, conn, number_of_objects, topic_events, create_threa
 
     # start lifecycle processing
     admin(['lc', 'process'], get_config_cluster())
-    print('wait for 20s for the lifecycle...')
-    time.sleep(20)
     print('wait for sometime for the messages...')
-
+    max_loops = 100
     no_keys = list(bucket.list())
+    while not no_keys and max_loops > 0:
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1797]:
```suggestion
    while len(no_keys) > 0 and max_loops > 0:
```

File: src/test/rgw/bucket_notification/test_bn.py [hunk 4]
```diff
@@ -1791,30 +1791,38 @@ def lifecycle(endpoint_type, conn, number_of_objects, topic_events, create_threa
...
+            for event_key in event_keys:
+                if event_key == key.name:
+                    key_found = True
+                    break
+            if not key_found:
+                err = 'no lifecycle event found for key: ' + str(key)
+                log.error(events)
+                assert False, err
+    else:
+        print('polling on bucket objects reached max timeout, aborting test!')
```
Comment (yuvalif) [inline src/test/rgw/bucket_notification/test_bn.py:1825]: you can assert and fail the test in this case

**Review by yuvalif — APPROVED** [review]

**Comment by AliMasarweh** [comment]
The PR will decrease the frequency of http failure, but we might see them from time to time ( failing tests 1to 8),
Passed jobs:
https://pulpito.ceph.com/alimasa-2024-07-30_13:34:22-rgw:notifications-wip-alimasa-bn-test-fix-http-distro-default-smithi
https://pulpito.ceph.com/alimasa-2024-07-30_13:34:35-rgw:notifications-wip-alimasa-bn-test-fix-http-distro-default-smithi
https://pulpito.ceph.com/alimasa-2024-07-30_13:34:49-rgw:notifications-wip-alimasa-bn-test-fix-http-distro-default-smithi
https://pulpito.ceph.com/alimasa-2024-07-30_14:20:29-rgw:notifications-wip-alimasa-bn-test-fix-http-distro-default-smithi
https://pulpito.ceph.com/alimasa-2024-07-30_14:20:40-rgw:notifications-wip-alimasa-bn-test-fix-http-distro-default-smithi
https://pulpito.ceph.com/alimasa-2024-07-30_14:20:46-rgw:notifications-wip-alimasa-bn-test-fix-http-distro-default-smithi
https://pulpito.ceph.com/alimasa-2024-07-30_14:20:53-rgw:notifications-wip-alimasa-bn-test-fix-http-distro-default-smithi
Failed:
https://pulpito.ceph.com/alimasa-2024-07-30_13:34:42-rgw:notifications-wip-alimasa-bn-test-fix-http-distro-default-smithi

**Comment by AliMasarweh** [comment]
jenkins test api

**Comment by AliMasarweh** [comment]
jenkins test windows

**Comment by AliMasarweh** [comment]
jenkins test api

**Comment by AliMasarweh** [comment]
jenkins test api

**Comment by AliMasarweh** [comment]
jenkins test api

**Comment by AliMasarweh** [comment]
jenkins test api

**Comment by AliMasarweh** [comment]
jenkins test api

**Pull request merged**